components/director/internal/open_resource_discovery/validation.go

```diff
@@ -937,3 +1060,24 @@ func validateExtensibleInnerFields(el gjson.Result) error {
 	}
 	return nil
 }
+
+func HashObject(obj interface{}) (uint64, error) {
+	hash, err := hashstructure.Hash(obj, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		return 0, errors.New("failed to hash the given object")
+	}
+
+	return hash, nil
+}
+
+func checkHashEquality(rdFromDBVersion, rdFromDocVersion, hashFromDB, hashFromDoc string) error {
+	rdFromDBVersion = fmt.Sprintf("v%s", rdFromDBVersion)
+	rdFromDocVersion = fmt.Sprintf("v%s", rdFromDocVersion)
+
+	areVersionsEqual := semver.Compare(rdFromDocVersion, rdFromDBVersion)
+	if areHashesEqual := cmp.Equal(hashFromDB, hashFromDoc); !areHashesEqual && areVersionsEqual <= 0 {
```

Comment:
Here a validation case is missed - if there are no changes to the resource but the version is changed (incremented, decremented). This should be illegal and an error should be thrown.

Example:

	areHashesEqual := cmp.Equal(hashFromDB, hashFromDoc)

	if areHashesEqual && areVersionsEqual != 0 {
		return errors.New("changed version without any resource changes")
	}

	if !areHashesEqual && areVersionsEqual <= 0 {
		return errors.New("there is a change in the resource; version value should be incremented")
	}